src/core/stream_send.c

```diff
@@ -1495,6 +1554,32 @@ QuicStreamOnAck(
     QuicStreamValidateRecoveryState(Stream);
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicStreamCleanupReliableReset(
+    _In_ QUIC_STREAM* Stream
+) {
+    //
+    // Cleanup. Cancels all send requests with offsets after ReliableOffsetSend.
+    // Assume this function gets called only when we're about to close the SEND
+    // Path of this stream once sufficient data has been received and ACK'd by the peer.
+    //
+    while (Stream->SendRequests) {
+        QUIC_SEND_REQUEST* Req = Stream->SendRequests;
+        //
+        // Ignore send requests that should have been ACK'd, i.e., < ReliableOffsetSend
+        //
+        if (Req->StreamOffset + Req->TotalLength < Stream->ReliableOffsetSend) {
+            continue;
+        }
+        Stream->SendRequests = Req->Next;
+        if (Stream->SendRequests == NULL) {
+            Stream->SendRequestsTail = &Stream->SendRequests;
+        }
+        QuicStreamCompleteSendRequest(Stream, Req, FALSE, TRUE);
```

Comment:
You must complete sends in the order they were queued. This seems to skip over requests still in the queue that are valid, and then cancel the ones after them. We can't do that. You have to wait to cancel these until the others have completed.

Ideally, we should have a test case specifically for this scenario; that requests are completed in the right order.

Reply:
Sorry! This bit of code was added while I was figuring out the design of my system. Notice how the code coverage never hits it, as this code only runs when UnAckedOffset >= Reliable, so it is guaranteed to complete send requests in order. I can get rid of the useless code that skips over the always false case.